File: historyscan-back/src/main/java/io/sch/historyscan/infrastructure/features/codebase/clone/AddedCodebaseDTO.java
package io.sch.historyscan.infrastructure.features.codebase.clone;

import com.fasterxml.jackson.annotation.JsonCreator;
import io.sch.historyscan.domain.error.HistoryScanTechnicalException;
import io.sch.historyscan.infrastructure.features.codebase.CodeBaseController;
import org.springframework.hateoas.RepresentationModel;
import org.springframework.http.HttpMethod;

import java.util.Objects;

import static org.springframework.hateoas.server.mvc.WebMvcLinkBuilder.linkTo;

public class AddedCodebaseDTO extends RepresentationModel<AddedCodebaseDTO> {

    private final String name;
    private final String url;
    private final String currentBranch;

    @JsonCreator
    public AddedCodebaseDTO(String name,
                            String url,
                            String currentBranch) {
        this.name = name;
        this.url = url;
        this.currentBranch = currentBranch;
        try {
            addSelfLink();
            codeBasesListLink();
        } catch (NoSuchMethodException e) {
            throw new HistoryScanTechnicalException("No method found in the controller", e);
        }
    }

    private void codeBasesListLink() throws NoSuchMethodException {
        add(linkTo(CodeBaseController.class,
                CodeBaseController.class.getMethod("currentCodeBases"))
                .withRel("codebases-list")
                .withTitle(HttpMethod.GET.name()));
    }

    private void addSelfLink() throws NoSuchMethodException {
        add(linkTo(CodeBaseController.class,
                CodeBaseController.class.getMethod("add", CodeBaseToAddDTO.class))
                .withSelfRel()
                .withTitle(HttpMethod.POST.name()));

    }

    public String getName() {
        return name;
    }

    public String getUrl() {
        return url;
    }

    public String getCurrentBranch() {
        return currentBranch;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        if (!super.equals(o)) return false;
        AddedCodebaseDTO that = (AddedCodebaseDTO) o;
        return Objects.equals(currentBranch, that.currentBranch)
                && Objects.equals(name, that.name)
                && Objects.equals(url, that.url);
    }

    @Override
    public int hashCode() {
        return Objects.hash(super.hashCode(), name, url, currentBranch);
    }
}

File: historyscan-back/src/main/kotlin/io/sch/historyscan/infrastructure/features/codebase/clone/AddedCodebaseDTO.kt
package io.sch.historyscan.infrastructure.features.codebase.clone

import com.fasterxml.jackson.annotation.JsonCreator
import io.sch.historyscan.domain.error.HistoryScanTechnicalException
import io.sch.historyscan.infrastructure.features.codebase.CodeBaseController
import org.springframework.hateoas.RepresentationModel
import org.springframework.hateoas.server.mvc.WebMvcLinkBuilder
import org.springframework.http.HttpMethod
import java.util.*

class AddedCodebaseDTO @JsonCreator constructor(val name: String,
                                                val url: String,
                                                val currentBranch: String) : RepresentationModel<AddedCodebaseDTO?>() {
    init {
        try {
            addSelfLink()
            codeBasesListLink()
        } catch (e: NoSuchMethodException) {
            throw HistoryScanTechnicalException("No method found in the controller", e)
        }
    }

    @Throws(NoSuchMethodException::class)
    private fun codeBasesListLink() {
        add(WebMvcLinkBuilder.linkTo(CodeBaseController::class.java,
                CodeBaseController::class.java.getMethod("currentCodeBases"))
                .withRel("codebases-list")
                .withTitle(HttpMethod.GET.name()))
    }

    @Throws(NoSuchMethodException::class)
    private fun addSelfLink() {
        add(WebMvcLinkBuilder.linkTo(CodeBaseController::class.java,
                CodeBaseController::class.java.getMethod("add", CodeBaseToAddDTO::class.java))
                .withSelfRel()
                .withTitle(HttpMethod.POST.name()))
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        if (!super.equals(o)) return false
        val that = o as AddedCodebaseDTO
        return currentBranch == that.currentBranch && name == that.name && url == that.url
    }

    override fun hashCode(): Int {
        return Objects.hash(super.hashCode(), name, url, currentBranch)
    }
}
